juju/juju
# [dev] refactor installation of lxd, kvm and mongo deps

## Description of change

This PR originally began as a tentative fix for the attached bug. The bug manifests itself when trying to create an lxd container on a host running **cosmic** or **disco** that does not come with lxd pre-installed. While all disco images seem to come with lxd pre-installed (as a snap vs an apt pkg for bionic and earlier), this is not the case with images for the s390 arch.  By default, we try to install lxd via **apt**. 

The apt repos contain a transition package that actually performs `snap install lxd`. However, the package which is currently available cannot be installed in non-interactive mode: it displays a prompt for the operator to select which lxd version to install. Unfortunately, the prompt cannot be bypassed causing apt-get to block forever.

The solution for this issue is to install via snap on cosmic and onwards which essentially means 
we need to resolve to series-sniffing to decide which package manager (and package names) to use (the juju/packaging repo now includes a snap-based package manager via https://github.com/juju/packaging/pull/4 and https://github.com/juju/packaging/pull/5)

However, as I went through the code base I saw that we use the same pattern in more places:
- to install mongo dependencies when bootstrapping controllers
- to install kvm dependencies. The latter only works on ubuntu and only happens to work across different series because the required package names have not been renamed.

Evidently, the old approach where we first consulted the series to select a list of packages and then asked the `juju/packaging` repo to give us a suitable package manager cannot cover systems where multiple package managers can co-exist (e.g. apt/snap). Additionally, there are cases where we would want to install some packages from snap and others from snap (i.e. they are not available as snaps yet, we don't want them to auto-update etc.).

The solution proposed by this PR is to create an abstraction for a `Dependency`. Each dependency contains the package-selection logic for the series that we support and can be queried to obtain a list of packages **and** preferred package manager for installing each one. I have isolated the dependency requirements for lxd, kvm and mongo into standalone objects that live in the `packaging/dependency` juju pkg. 

The `packaging` package provides a helper for installing dependencies and handles all the low-level details for instantiating the appropriate package managers and injecting additional options (e.g. backports, cloud-tools) where required.

Finally, as an application of the boy-scout rule, I noticed that the `mongo` package contains centos-specific code (and tests) for bootstrapping purposes. Given that we only support bootstrapping on ubuntu (centos is fine for running workloads though), that code is never going to be used so it has been removed. As an extra layer of safety, the dependency object for mongo will return a `NotImplemented` error if asked to resolve the mongo packages on centos.

**Notes:**
To reduce the scope of changes for this PR, I have intentionally left out the changes required to configure proxies for systems with dual package managers (e.g. snap/apt on disco). This will be addressed via a separate PR

## QA steps

To simulate an image where lxd is not pre-installed, we will add two machines (based on bionic and disco), uninstall lxd and then try to add lxd containers on each. This should force a re-install of lxd using he appropriate package manager for each distro.

```console
$ juju bootstrap lxd test --no-gui
$ juju add-machine --series=bionic
$ juju add-machine --series=disco

# Wait for the machines to become available and run the following commands
$ juju run --machine 0 'sudo apt-get --option=Dpkg::Options::=--force-confold --option=Dpkg::options::=--force-unsafe-io --assume-yes --quiet purge lxd'
...
Removing lxd (3.0.3-0ubuntu1~18.04.1) ...
...

$ juju run --machine 1 'sudo snap remove lxd'
lxd removed

# Now create two lxd containers and check that juju re-install lxd using the correct package manager
$ juju add-machine lxd:0
created container 0/lxd/0

$ juju add-machine lxd:1
created container 1/lxd/0
 
# Wait for containers to become available and check lxd installations on host machines
$ juju run --machine 0 'which lxd'
/usr/bin/lxd       <-- apt version on bionic host

$ juju run --machine 1 'which lxd'
/snap/bin/lxd      <-- snap version on disco host
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1835187
